[WIP] Handler context

danielmarbach opened the pull request:
- [x] Define ATTs for old and new style handler interop, message handler ordering, sagas... (by @danielmarbach)
- [x] Design and implement HandlerOrdering, Filtering based on new interfaces and also handler invocation
- [x] Design, discuss and implement `IAmStartedByYouNameIt`
- [x]  Design, discuss and implement `IHandleTimeoutYouNameIt`
- [x] Design, discuss and implement saga base class and/or need for reflection checks about usage of new APIs on saga (more questions can be found [here](https://github.com/Particular/NServiceBus/commit/b5d03d92228391d24b3d857181fb834f10805a84#diff-273ca941a7027d949f9abe00ec4130dcR44)
https://github.com/Particular/NServiceBus/commit/47c161da195e9a2505354e068a5d5590c7250f4d)
- [x] Add new behaviors which execute the new style API
- [x] Duplicate existing ATTs and rewrite to new APIs where necessary (@danielmarbach)
- [x] Design, discuss and implement contexts (no inheritance, currently aggregate IBus under the cover)

danielmarbach commented:
Currently `LoadHandlersConnector` loads the handler types and creates the necessary handler invocation objects and dispatches it to the next stage. I think introducing a new stage for the new behaviors doesn't make sense. My current approach is to treat the different consumer interfaces exactly the same in the core to got the ball rolling. Should we later differentiate `IHandle` (working title) and `ISubscribe` (working title) differently in the core (like having `MessageHandlerRegistry`, `SubscribeHandlerRegistry`...)?

danielmarbach commented:
@johnsimons I defined a few ATTs as guidance to start implementing the feature. The revealed a bunch of questions. Can you have a look at the ATTs? Should we also write one which shows all the mutators and inspect whether we need to change those as well?

johnsimons commented on src/NServiceBus.Core/Saga/IAmStartedByMessages.cs:
```diff
@@ -9,4 +9,16 @@ namespace NServiceBus.Saga
+
+#pragma warning disable 1591
+
+    // TODO: We need to discuss whether this makes sense. This is just a first shot.
```
I never liked those interfaces, IMO they are too confusing.

johnsimons commented on src/NServiceBus.Core/Saga/IHandleTimeouts.cs:
```diff
@@ -10,4 +10,14 @@ public interface IHandleTimeouts<T>
         void Timeout(T state);
     }
+#pragma warning disable 1591
+    // Daniel :This is not thought through and serves as a discussion point. Should we have timeout context. If yes should it contain the state and contextual information including contextual bus.
```
:+1:

johnsimons commented on src/NServiceBus.AcceptanceTests/HandlerContext/When_saga_with_old_and_new_api_is_created.cs:
```diff
@@ -0,0 +1,115 @@
+                EndpointSetup<DefaultPublisher>();
+            }
+
+            public class Saga1 : Saga<Saga1.Saga1Data>, // Daniel: I think we need a new base class which doesn't have the IBus already predefined or should we use the same base and advice users to use the context information?
```
:+1: to have a new base

johnsimons commented on src/NServiceBus.AcceptanceTests/HandlerContext/When_saga_with_old_and_new_api_is_created.cs:
```diff
@@ -0,0 +1,115 @@
+            public class Saga1 : Saga<Saga1.Saga1Data>, // Daniel: I think we need a new base class which doesn't have the IBus already predefined or should we use the same base and advice users to use the context information?
+                IAmStartedByMessages<StartSaga>, // Daniel: I think we shouldn't allow this. But technically it is possible
+                IAmStartedByMessage<StartSaga>, // Daniel: see above you need to pick and choose
+                IAmStartedByEvent<SomethingHappenedEvent>, // Daniel: I'm started by event?
```
Looking at all the IAmStartedBy... it feels weird.
I wish we removed all those and instead define what messages start a saga as part of the mappings, thoughts ?
/cc @Particular/nservicebus

andreasohlund replied:
Would definitely help if we ever want POCO sagas

danielmarbach replied:
Or turn it around. Because in practice there is rarely a case where a
message should not start a saga why not make it by default so that a
message can start a saga and you have to define which messages do not start
a saga?

johnsimons replied:
That could work, I always been worried about this api, I wonder what
happens if there is no istartsaga at all in a saga, I guess it never starts!

On Tuesday, March 3, 2015, Daniel Marbach <notifications@github.com> wrote:

> In
> src/NServiceBus.AcceptanceTests/HandlerContext/When_saga_with_old_and_new_api_is_created.cs
> <https://github.com/Particular/NServiceBus/pull/2669#discussion_r25678807>
> :
>
> > +        {
> > +            public bool SagaCompleted { get; set; }
> > +            public bool SagaStarted { get; set; }
> > +        }
> > +
> > +        public class SagaEndpoint : EndpointConfigurationBuilder
> > +        {
> > +            public SagaEndpoint()
> > +            {
> > +                EndpointSetup<DefaultPublisher>();
> > +            }
> > +
> > +            public class Saga1 : Saga<Saga1.Saga1Data>, // Daniel: I think we need a new base class which doesn't have the IBus already predefined or should we use the same base and advice users to use the context information?
> > +                IAmStartedByMessages<StartSaga>, // Daniel: I think we shouldn't allow this. But technically it is possible
> > +                IAmStartedByMessage<StartSaga>, // Daniel: see above you need to pick and choose
> > +                IAmStartedByEvent<SomethingHappenedEvent>, // Daniel: I'm started by event?
>
> Or turn it around. Because in practice there is rarely a case where a
> message should not start a saga why not make it by default so that a
> message can start a saga and you have to define which messages do not start
> a saga?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/2669/files#r25678807>.
>

danielmarbach replied:
@Particular/nservicebus 

How do you like this idea:

[see](https://github.com/Particular/NServiceBus/pull/2669/files#r25678807)

danielmarbach commented:
Got stuck in the afternoon how to approach #2680

danielmarbach commented:
@Particular/nservicebus I think I fixed now the handler instantiation and invocation. I moved that responsibility into the MessageHandler decorator class. I still need to cleanup the registry and the handler class and also reenable the HandlerInvocationCache tests. The cool thing is that each IHandle, IHandleMessages, ISubscribe etc. is treated as a handler. Feel free to comment (although still work todo, i.ex. get rid of the NullContext)

SeanFeldman commented on src/NServiceBus.Core.Tests/API/approvals/NServiceBus.Core.approved.cs:
```diff
@@ -439,6 +461,11 @@ public interface IStartableBus : NServiceBus.IBus, NServiceBus.ISendOnlyBus, Sys
     {
         NServiceBus.IBus Start();
     }
+    public interface ISubscribe<T>
```
Name suggestion: `ISubscribeTo<T>`

indualagarsamy commented on src/NServiceBus.Core/Unicast/Behaviors/MessageHandler.cs:
```diff
@@ -7,14 +7,76 @@
+    [Flags]
+    internal enum HandlerKind
+    {
+        None = 0,
```
Do we need 'None' ?

indualagarsamy commented on src/NServiceBus.Core/Unicast/Behaviors/MessageHandler.cs:
```diff
@@ -7,14 +7,76 @@
+    /// <summary>
+    /// Daniel: Should we expose that to public?
+    /// </summary>
+    [Flags]
```
Do we need flags?

danielmarbach commented:
Started implementing checks for illegal usage of new interfaces on sagas. My current thinking is to do that inside the metamodel. Any objections?

danielmarbach commented:
@andreasohlund suggested to use `IConsumeMessage`, `IConsumeEvent` and `IConsumeTimeout`.

danielmarbach commented:
@johnsimons I added a few validation tests for the TypeBasedSagaMetaModel. I'm still thinking about a good validation strategy to test the combination of allowed interfaces. Any good ideas? Because this doesn't work:

````
       static IEnumerable<SagaMessage> GetAssociatedMessages(Type sagaType)
        {
            var result = GetMessagesCorrespondingToFilterOnSaga(sagaType, typeof(IAmStartedByMessages<>))
                .Select(t => new SagaMessage(t.FullName, true)).ToList();

            foreach (var messageType in GetMessagesCorrespondingToFilterOnSaga(sagaType, typeof(IAmStartedByMessage<>)))
            {
                if (result.Any(m => m.MessageType == messageType.FullName))
                {
                    var oldType = typeof(IAmStartedByMessages<>).MakeGenericType(messageType);
                    var newType = typeof(IAmStartedByMessage<>).MakeGenericType(messageType);
                    throw new Exception(string.Format("Saga {0} cannot implement {1} and {2} for the same message type {3} at the same time. Consider removing {1}.", sagaType.FullName, oldType.FullName, newType.FullName, messageType.FullName));
                }
                result.Add(new SagaMessage(messageType.FullName, true));
            }

            foreach (var messageType in GetMessagesCorrespondingToFilterOnSaga(sagaType, typeof(IAmStartedByEvent<>)))
            {
                if (result.Any(m => m.MessageType == messageType.FullName))
                {
                    var oldType = typeof(IAmStartedByMessages<>).MakeGenericType(messageType);
                    var newType = typeof(IAmStartedByEvent<>).MakeGenericType(messageType);
                    throw new Exception(string.Format("Saga {0} cannot implement {1} and {2} for the same message type {3} at the same time. Consider removing {1}.", sagaType.FullName, oldType.FullName, newType.FullName, messageType.FullName));
                }
                result.Add(new SagaMessage(messageType.FullName, true));
            }

            foreach (var messageType in GetMessagesCorrespondingToFilterOnSaga(sagaType, typeof(IHandleMessages<>)))
            {
                if (result.Any(m => m.MessageType == messageType.FullName))
                {
                    continue;
                }
                result.Add(new SagaMessage(messageType.FullName, false));
            }

            foreach (var messageType in GetMessagesCorrespondingToFilterOnSaga(sagaType, typeof(IHandle<>)))
            {
                if (result.Any(m => m.MessageType == messageType.FullName))
                {
                    var oldType = typeof(IHandleMessages<>).MakeGenericType(messageType);
                    var newType = typeof(IHandle<>).MakeGenericType(messageType);
                    throw new Exception(string.Format("Saga {0} cannot implement {1} and {2} for the same message type {3} at the same time. Consider removing {1}.", sagaType.FullName, oldType.FullName, newType.FullName, messageType.FullName));
                }
                result.Add(new SagaMessage(messageType.FullName, false));
            }

            foreach (var messageType in GetMessagesCorrespondingToFilterOnSaga(sagaType, typeof(ISubscribe<>)))
            {
                if (result.Any(m => m.MessageType == messageType.FullName))
                {
                    var oldType = typeof(IHandleMessages<>).MakeGenericType(messageType);
                    var newType = typeof(ISubscribe<>).MakeGenericType(messageType);
                    throw new Exception(string.Format("Saga {0} cannot implement {1} and {2} for the same message type {3} at the same time. Consider removing {1}.", sagaType.FullName, oldType.FullName, newType.FullName, messageType.FullName));
                }
                result.Add(new SagaMessage(messageType.FullName, false));
            }

            foreach (var messageType in GetMessagesCorrespondingToFilterOnSaga(sagaType, typeof(IHandleTimeouts<>)))
            {
                if (result.Any(m => m.MessageType == messageType.FullName))
                {
                    continue;
                }
                result.Add(new SagaMessage(messageType.FullName, false));
            }

            foreach (var messageType in GetMessagesCorrespondingToFilterOnSaga(sagaType, typeof(IHandleTimeout<>)))
            {
                if (result.Any(m => m.MessageType == messageType.FullName))
                {
                    var oldType = typeof(IHandleTimeouts<>).MakeGenericType(messageType);
                    var newType = typeof(IHandleTimeout<>).MakeGenericType(messageType);
                    throw new Exception(string.Format("Saga {0} cannot implement {1} and {2} for the same message type {3} at the same time. Consider removing {1}.", sagaType.FullName, oldType.FullName, newType.FullName, messageType.FullName));
                }
                result.Add(new SagaMessage(messageType.FullName, false));
            }

            return result;
        }
````

danielmarbach commented:
@johnsimons can you give some thought what would be the minimum amount of methods we would require on the different context implementations? 

Furthermore how should the mutator context look like?

johnsimons commented:
@danielmarbach I'll start reviewing this today

danielmarbach commented:
Based on the input of @andreasohlund @johnsimons and I discussed today if we should have also a `IConsumeMessage`, `IConsumeCommand`, `IConsumeEvent` and `IConsumeTimeout`.

Currently in the handler context we differentiate between events and command|messages. From a high level perspective we also different commands from messages (conventions do but also the interfaces `ICommand`, `IMessage`, `IEvent`. Having a seperate consumer interface would allow us to guide the users more in the future. Or is that overkill?

@udidahan @Particular/engineers thoughts?

SzymonPobiega commented:
How about the `IcomsumeMessageIdempotently` in order to make it apparent for the user that in some modes he has to take care of idempotency himself?

danielmarbach commented:
I wouldn’t put  that into the name of the interface.

SzymonPobiega commented:
True. It can be better addressed using an attribute in a separate feature. Scratch that.

danielmarbach commented:
Based on https://github.com/Particular/NServiceBus/pull/2669#issuecomment-85921038 I pushed now the following new interfaces

* `IConsumeMessage`
* `IConsumeEvent`
* `IConsumeTimeout`
* `IAmStartedByConsumedMessage`
* `IAmStartedByConsumedEvent`

I still think we should have 
* `IConsumeCommand`
* `IAmStartedByConsumedCommand`

for better differentiation between messages and commands.

Furthermore I have now validation logic which checks sagas for invalid combinations of both old and new style API usages. While implementing I came to the conclusion that following should not be done:

```
class SagaWithHandleEventOldAndNew : Saga<SagaWithHandleEventOldAndNew.SagaData>,
           IHandleMessages<SagaWithHandleEventOldAndNew.Event>,
           IConsumeEvent<SagaWithHandleEventOldAndNew.Event>

        class SagaWithHandleMessageOldAndNew : Saga<SagaWithHandleMessageOldAndNew.SagaData>,
            IHandleMessages<SagaWithHandleMessageOldAndNew.StartSaga>,
            IConsumeMessage<SagaWithHandleMessageOldAndNew.StartSaga>
        
class SagaWithConsumeTimeoutOldAndNew : 
Saga<SagaWithConsumeTimeoutOldAndNew.SagaData>,
            IHandleTimeouts<SagaWithConsumeTimeoutOldAndNew.TimeoutData>,
            IConsumeTimeout<SagaWithConsumeTimeoutOldAndNew.TimeoutData>

        class SagaWithIAmStartedWithConsumedEventOldAndNew : Saga<SagaWithIAmStartedWithConsumedEventOldAndNew.SagaData>,
            IAmStartedByMessages<SagaWithIAmStartedWithConsumedEventOldAndNew.Event>,
            IAmStartedByConsumedEvent<SagaWithIAmStartedWithConsumedEventOldAndNew.Event>

        class SagaWithIAmStartedOldAndNew : Saga<SagaWithIAmStartedOldAndNew.SagaData>,
            IAmStartedByMessages<SagaWithIAmStartedOldAndNew.StartSaga>,
            IAmStartedByConsumedMessage<SagaWithIAmStartedOldAndNew.StartSaga>
```

Anything I forgot?

I quickly reviewed the validation logic with @janovesk and discussed perf implications about the Any() usages. We came to the conclusion it is OK for saga validation

udidahan commented:
I prefer IProcessMessages to IConsumeMessages.
Also - why the singular instead of the plural: Message rather than Messages?

danielmarbach commented:
My question first ;) : what about seperating commands from messages?

The idea behind the singular was that from a handler perspective you only
consume one message at a time. But I'm not really opinionated about that.
Change it to plural?

Am Donnerstag, 26. März 2015 schrieb Udi Dahan :

> I prefer IProcessMessages to IConsumeMessages.
> Also - why the singular instead of the plural: Message rather than
> Messages?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/2669#issuecomment-86606635>
> .
>


-- 

Daniel Marbach

Solution Architect, Particular Software

Email:  daniel.marbach@particular.net

Web:    www.particular.net
<http://t.signauxneuf.com/e1t/c/5/f18dQhb0S7lC8dDMPbW2n0x6l2B9nMJN7t5XX4RzDjlVS9QB88pTc_dW3Ljg7x56dWyBdgBXs402?t=http%3A%2F%2Fwww.particular.net%2F&si=4896787407044608&pi=33cc6202-1f31-4040-8d6c-e70641f44bba>

udidahan commented:
@danielmarbach How about reframing the question as separating commands from responses - it wouldn't make sense to respond to a response. In which case, if we create interfaces for commands, events, and responses - why would we need one for generic messages?

danielmarbach commented:
Good for me. This is a total different approach. It would mean RIP (in the
long run) for IMessage and DefineMessages convention and introduce
IResponse as well as DefineResponse convention. I like it from thw
differentiation point of view but it also implicates more changes for the
customers. What's your thought?

-- 

Daniel Marbach

Solution Architect, Particular Software

Email:  daniel.marbach@particular.net

Web:    www.particular.net
<http://t.signauxneuf.com/e1t/c/5/f18dQhb0S7lC8dDMPbW2n0x6l2B9nMJN7t5XX4RzDjlVS9QB88pTc_dW3Ljg7x56dWyBdgBXs402?t=http%3A%2F%2Fwww.particular.net%2F&si=4896787407044608&pi=33cc6202-1f31-4040-8d6c-e70641f44bba>

udidahan commented:
See what the rest of the techies have to say. What do I know - I'm just a CEO :)

johnsimons commented:
:+1: for singular names

danielmarbach commented:
Introducing command processors and leaving message processors is a little
less changes in one Go than also introducing IResponse. But you could argue
that the handler context changed in the long run are significant for the
users anyway and therefore making additional changes from IMessage to
IResponse doesn't hurt more. Not sure what is better. See what others think



-- 

Daniel Marbach

Solution Architect, Particular Software

Email:  daniel.marbach@particular.net

Web:    www.particular.net
<http://t.signauxneuf.com/e1t/c/5/f18dQhb0S7lC8dDMPbW2n0x6l2B9nMJN7t5XX4RzDjlVS9QB88pTc_dW3Ljg7x56dWyBdgBXs402?t=http%3A%2F%2Fwww.particular.net%2F&si=4896787407044608&pi=33cc6202-1f31-4040-8d6c-e70641f44bba>

johnsimons commented:
I think it is the right time to rename IMessage to IResponse

andreasohlund commented:
>Change it to plural?

+1 for changing since I believe it reads better: `IConsumeMessages<T>` => "I consume messages of type T" instead of "I consume a single message of type T"

andreasohlund commented:
>I think it is the right time to rename IMessage to IResponse

Not against this (but I guess we would deprecate IMessage and introduce a new IResponse since they are not the same thing)

Since we've always talked about making the conventions optional will this make that harder? (I don't think so)

For example (with no conventions or marker interfaces)

`bus.Send(new MyMessage())` => Will classify MyMessage as a command

`bus.Reply(new MyResponse())` => Will classify MyResponseas a response

`bus.Publish(new MyEvent())` => Will classify MyEvent as a event

So if first one win this would throw:

``` 

bus.Send(new MyMessage())
bus.Publish(new MyMessage()) // throw since MyMessage is a command

```

Thoughts?

danielmarbach commented:
So let's do a combination:

* `IProcessCommands`
* `IProcessResponses`
* `IProcessEvents`
* `IProcessTimeouts`
* `IAmStartedByProcessedCommands`
* `IAmStartedByProcessedResponses`
* `IAmStartedByProcessedEvents`

@udidahan @johnsimons @andreasohlund @SzymonPobiega ok?

SzymonPobiega commented:
This would improve our support for command/event messages as well as make the request/response pattern explicit but on the other hand it would make it harder to implement [document style](http://www.enterpriseintegrationpatterns.com/DocumentMessage.html) messaging. `IMessage` is not ideal but would get the job done. I don't think a command can be used for document style since I believe we should validate there is a single handler for any given command.

Raising a separate issue https://github.com/Particular/Requirements/issues/257 for document style messaging.

SzymonPobiega commented:
After talking it through with @danielmarbach I am good with removing `IMessage`
:+1:

udidahan commented:
I don't like IAmStartedByProcessedCommands - I'd rather have IAmStartedByCommands.

Also, I'm not sure that there would be any circumstances in which a response would start a saga - do you?

Finally, I'm not suggested removing IMessage - I think that needs to stick around, but without any of this new contextual goodness.

danielmarbach commented:
@udidahan 
> Also, I'm not sure that there would be any circumstances in which a response would start a saga - do you?
Me neither. But technically it is possible now with `IAmStartedByMessages`. So you are saying we shouldn't support that in the next generation APIs.

> I don't like IAmStartedByProcessedCommands - I'd rather have IAmStartedByCommands.
Ok. And then `IAmStartedByEvents` as well I assume

> Finally, I'm not suggested removing IMessage - I think that needs to stick around, but without any of this new contextual goodness.
To be clear we are not removing anything yet. As the plan of attack mentioned we will support both old and new style APIs side by side for some time

danielmarbach commented:
I'm using now `IReply` instead of `IResponse`. Because Reply is the better term than response.

danielmarbach commented:
Just pushed the new API proposal.

udidahan commented:
> And then IAmStartedByEvents as well I assume

Yes.

> But technically it is possible now with IAmStartedByMessages.

Well, since we didn't have the command/event specific interfaces - it was the only way of writing sagas. That doesn't mean that it was used to start sagas from responses/replies.

> IProcessReplies is up and running

I like ```IProcessReplies``` but am not thrilled with ```IReply``` - it sounds like it would be the object you'd use to send a reply. For that reason I think that ```IResponse``` is better, and then, by proxy, so would be ```IProcessResponses```.

danielmarbach commented:
@udidahan changed the name

johnsimons commented:
Still think the singular names are the way to go.
Also not really keen on ` IAmStartedBy...`, I would rather define what starts a saga as part of the saga config.

SzymonPobiega commented:
Shouldn't we assume any message can start a saga?

johnsimons commented:
Exactly my point @SzymonPobiega

danielmarbach commented:
@szymonpobiega and @johnsimons yes we should. We decided to NOT include
that in this PR. Ok?

johnsimons commented:
Not really @danielmarbach.
This PR is proposing to introduce 3 new interfaces, I would rather not and do it in a better way instead.

danielmarbach commented:
@johnsimons now we are running in circles. Have you been part of this meeting as well were we discussed the sagas messages to be auto starting and an opt-out configuration. We decided to not include that in this PR and possibly not even in V6 because of the risk. Therefore we need the new interfaces.

@andreasohlund thoughts?

johnsimons commented:
@danielmarbach I understand the need for the 3 new interfaces however taking a step back and looking at the proposal, IMHO it is way too many interfaces for a user to assimilate, we already have issues explaining what a saga is and how to take advantage of sagas, we are now proposing to make this even harder for the users, I would instead prefer to try to make it simpler.
Anyway just my view on this subject.

danielmarbach commented:
@johnsimons I do agree with you 100%. The question is should we bundle
those two changes together at all costs? Or can we finish that PR, merge it
and push through the necessary changes on the sagas?

andreasohlund commented:
+1 for separating it. Including the saga rewamp will make this one to big
IMO

On Tue, Mar 31, 2015 at 9:55 AM, Daniel Marbach <notifications@github.com>
wrote:

> @johnsimons I do agree with you 100%. The question is should we bundle
> those two changes together at all costs? Or can we finish that PR, merge it
> and push through the necessary changes on the sagas?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/2669#issuecomment-87986532>
> .
>

danielmarbach commented:
@andreasohlund right I think so as well. So this means in order to complete and merge this PR we need those interfaces.

danielmarbach commented:
@johnsimons @andreasohlund @SzymonPobiega which of those interfaces needs also a contextual overload (new style API)?

- [ ] NServiceBus.IResponse
- [ ] NServiceBus.IManageMessageHeaders
- [ ] NServiceBus.INeedInitialization
- [ ] NServiceBus.ISendOnlyBus
- [ ] NServiceBus.IDataBusProperty
- [ ] NServiceBus .IWantToRunBeforeConfigurationIsFinalize d
- [ ] NServiceBus .ISpecifyMessageHandlerOrdering
- [ ] NServiceBus.IAuthorizeSubscriptions
- [ ] NServiceBus.IBus
- [ ] NServiceBus.ICallback
- [ ] NServiceBus.IMessage
- [ ] NServiceBus.ICommand
- [ ] NServiceBus.IEvent
- [ ] NServiceBus.IMessageContext
- [ ] NServiceBus.IMessageCreator
- [ ] NServiceBus.IHandleMessages<T>
- [ ] NServiceBus.IProcessEvents<T>
- [ ] NServiceBus.IEventContext
- [ ] NServiceBus.IProcessCommands<T>
- [ ] NServiceBus.ICommandContext
- [ ] NServiceBus.IProcessResponses<T>
- [ ] NServiceBus.IResponseContext
- [ ] NServiceBus.IStartableBus
- [ ] NServiceBus .IWantToRunWhenBusStartsAndStops
- [ ] NServiceBus.IRunWhenBusStartsAndStops
- [ ] NServiceBus.IRunContext
- [ ] NServiceBus.IConfigureThisEndpoint
- [ ] NServiceBus.IExcludesBuilder
- [ ] NServiceBus.IIncludesBuilder
- [ ] NServiceBus.DataBus.IDataBusSerializer
- [ ] NServiceBus.DataBus.IDataBus
- [ ] NServiceBus.Logging.ILog
- [ ] NServiceBus.Logging.ILoggerFactory
- [ ] NServiceBus.Saga .IConfigureHowToFindSagaWithMessage
- [ ] NServiceBus.Saga.IContainSagaData
- [ ] NServiceBus.Saga.IAmStartedByMessages<T>
- [ ] NServiceBus.Saga.IAmStartedByCommands<T>
- [ ] NServiceBus.Saga.IAmStartedByEvents<T>
- [ ] NServiceBus.Saga.IFinder
- [ ] NServiceBus.Saga.IFindSagas<T>+Using<M>
- [ ] NServiceBus.Saga.IHandleSagaNotFound
- [ ] NServiceBus.Saga.IProcessSagaNotFound
- [ ] NServiceBus.Saga.ISagaNotFoundContext
- [ ] NServiceBus.Saga.IHandleTimeouts<T>
- [ ] NServiceBus.Saga.IProcessTimeouts<T>
- [ ] NServiceBus.Saga.ITimeoutContext
- [ ] NServiceBus.Saga.ISagaPersister
- [ ] NServiceBus.Serialization .IMessageSerializer
- [ ] NServiceBus.Satellites.ISatellite
- [ ] NServiceBus.Satellites .IAdvancedSatellite
- [ ] NServiceBus.Installation .INeedToInstallSomething
- [ ] NServiceBus.MessageMutator .IMutateOutgoingMessages
- [ ] NServiceBus.MessageMutator .IMutateIncomingMessages
- [ ] NServiceBus.MessageMutator .IMessageMutator
- [ ] NServiceBus.MessageMutator .IMutateOutgoingMessage
- [ ] NServiceBus.MessageMutator .IMutateIncomingMessage
- [ ] NServiceBus.MessageMutator .IMessagesMutator
- [ ] NServiceBus.MessageMutator .IMutateIncomingMessageContext
- [ ] NServiceBus.MessageMutator .IMutateIncomingTransportMessages
- [ ] NServiceBus.MessageMutator .IMutateIncomingTransportMessage
- [ ] NServiceBus.MessageMutator .IMutateIncomingTransportMessageContext
- [ ] NServiceBus.MessageMutator .IMutateOutgoingMessageContext
- [ ] NServiceBus.MessageMutator .IMutateOutgoingTransportMessages
- [ ] NServiceBus.MessageMutator .IMutateOutgoingTransportMessage
- [ ] NServiceBus.MessageMutator .IMutateOutgoingTransportMessageContext
- [ ] NServiceBus.MessageMutator .IMutateTransportMessages
- [ ] NServiceBus.MessageMutator .IMutateTransportMessage
- [ ] NServiceBus.Outbox.IOutboxStorage
- [ ] NServiceBus.UnitOfWork .IManageUnitsOfWork
- [ ] NServiceBus.Config .IWantToRunWhenConfigurationIsComplete
- [ ] NServiceBus.Config.ConfigurationSource .IConfigurationSource
- [ ] NServiceBus.Config.ConfigurationSource .IProvideConfiguration<T>
- [ ] NServiceBus.Pipeline.IBehavior<TIn,TOut>
- [ ] NServiceBus.Pipeline.IStageConnector
- [ ] NServiceBus.Pipeline.IExecutor
- [ ] NServiceBus.Pipeline.PipelineFactory
- [ ] NServiceBus.Encryption .IEncryptionService
- [ ] NServiceBus.Faults .IManageMessageFailures
- [ ] NServiceBus.Settings.ReadOnlySettings
- [ ] NServiceBus.Gateway.Deduplication .IDeduplicateMessages
- [ ] NServiceBus.MessageInterfaces .IMessageMapper
- [ ] NServiceBus.Unicast .IMessageHandlerRegistry
- [ ] NServiceBus.Unicast.Subscriptions .MessageDrivenSubscriptions .ISubscriptionStorage
- [ ] NServiceBus.Unicast.Queuing .IWantQueueCreated
- [ ] NServiceBus.ObjectBuilder.Common .IContainer
- [ ] NServiceBus.Timeout.Core .IPersistTimeouts

danielmarbach commented:
@SzymonPobiega something is off with the build. I don't get why the build is failing. When I look into the build server the merge is correct, even when I download the dependent artifacts and decompile it it contains the correct code which is green on my machine. Can you have a look what's wrong? This is driving me crazy

danielmarbach commented:
@Particular/engineers any ideas how to name the new interfaces?

johnsimons commented:
@danielmarbach have u consider just having 2 abstract classes for the mutators, one for the transport and another for the logical message ?

udidahan commented:
Why do we need to make changes to the mutators as a part of this requirement?

danielmarbach commented:
We discussed where else users are using bus and wanted to add contextual
bus there as well because the same argumentation we use for
handlers applies in mutators as well. Furthermore if we want to guide the
users to use our contextual buses at some point we need to deprecatr IBus.
The sooner they start using contextual buses the sooner we can do that

-- 

Daniel Marbach

Solution Architect, Particular Software

Email:  daniel.marbach@particular.net

Web:    www.particular.net
<http://t.signauxneuf.com/e1t/c/5/f18dQhb0S7lC8dDMPbW2n0x6l2B9nMJN7t5XX4RzDjlVS9QB88pTc_dW3Ljg7x56dWyBdgBXs402?t=http%3A%2F%2Fwww.particular.net%2F&si=4896787407044608&pi=33cc6202-1f31-4040-8d6c-e70641f44bba>

andreasohlund commented:
I'm on the fence here, we definitely want to avoid scope creep. But in the same way try create a consistent api and act on low hanging fruit as we come across them. Since there seems to be alot of discussion on the mutators perhaps it's a sign that it wasn't a low hanging fruit after all?

danielmarbach commented:
@udidahan I moved some of the commits to #2705 and also the list of interfaces

johnsimons commented:
The price to pay for evolving a 4+ year old API

danielmarbach commented:
@johnsimons I will finish the work on the ATTs tomorrow. If you start working on handler context and have any questions just ping me. I'm here for you buddy :D

danielmarbach commented:
@johnsimons I merged now the saga meta model changes into handler context. It forced me to expose [`SagaMessageHandledBy`](https://github.com/Particular/NServiceBus/commit/ef985c1783e3356ccfb2269ba9c4a1e3a530ebfc#diff-1eb180a4f6d35cd0c53fde55c1a077b3R1810). I'm still unsure about this. What is your thought on that?

danielmarbach commented:
@udidahan @andreasohlund @johnsimons @SzymonPobiega 
Should we allow to reply to an event published?

SzymonPobiega commented:
Replying to an event seems to be valid when doing anycast. Publish a message for anybody interested and wait for any response (first I guess) that comes. But this seems to be a very very specific use case.

udidahan commented:
We could say that to reply to an event, they would have to take a dependency on IBus.

danielmarbach commented:
@udidahan This approach works OK in the short run. But in the long run when we want to move users to the new APIs we need to deprecate IBus because only when we deprecate IBus we bring the users into the direction to remove any existing property or constructor dependency to IBus and therefore only then we can deprecate the old handling interfaces in the long run. So what would be the long term implementation here?

udidahan commented:
I don't believe we'll ever be able to deprecate it. People are doing vastly different things using very different approaches.

johnsimons commented:
> I don't believe we'll ever be able to deprecate it. People are doing vastly different things using very different approaches.

@udidahan it seems we are a bit misaligned where we want to take the NServiceBus API, should I prepare a presentation where we believe the NServiceBus API should be heading and how we see the API evolving in the coming years ?

danielmarbach commented:
@johnsimons right. I think we need to talk. Depending of the outcome I'd
start questioning the usefulness of this PR

Am Donnerstag, 9. April 2015 schrieb John Simons :

> I don't believe we'll ever be able to deprecate it. People are doing
> vastly different things using very different approaches.
>
> @udidahan <https://github.com/udidahan> it seems we are a bit misaligned
> where we want to take the NServiceBus API, should I prepare a presentation
> where we believe the NServiceBus API should be heading and how we see the
> API evolving in the coming years ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/2669#issuecomment-91064969>
> .
>


-- 

Daniel Marbach

Solution Architect, Particular Software

Email:  daniel.marbach@particular.net

Web:    www.particular.net
<http://t.signauxneuf.com/e1t/c/5/f18dQhb0S7lC8dDMPbW2n0x6l2B9nMJN7t5XX4RzDjlVS9QB88pTc_dW3Ljg7x56dWyBdgBXs402?t=http%3A%2F%2Fwww.particular.net%2F&si=4896787407044608&pi=33cc6202-1f31-4040-8d6c-e70641f44bba>

udidahan commented:
@johnsimons yes, please do.

danielmarbach commented:
@johnsimons Let me help!

johnsimons commented:
ok @udidahan, we will prepare a presentation for you, when we ready we will ping @Chante-Vaturi to book a time

johnsimons commented:
@danielmarbach not sure about the `SagaMessageHandledBy` enum, to me a message is of type timeout, command, message or event and it has associated extra metadata with it about if that message is allowed to start a new saga or not.
I wouldn't combine both into a single enum.

johnsimons commented:
Also not sure about making `SagaMessageHandledBy` public

danielmarbach commented:
@johnsimons collapsing would work because currently I don't really need the StartedBy... flags. Go ahead if you feel so

johnsimons commented:
@danielmarbach you are using StartedBy... flags

danielmarbach commented:
@johnsimons let me look at the code again. It is early morning after the weekend. Be patient with me :dart:

danielmarbach commented:
@johnsimons Make it internal again?

johnsimons commented:
:+1: @danielmarbach

danielmarbach commented:
@andreasohlund I would like to pull this into develop and then start working on the handler context interfaces and also the API on top of this interfaces. But this PR is becoming large enough I'd say. Thoughts?

The state is that we have now context objects without inheritance which aggregate `IBus` underneath. It still supports Replying in event scenarios. What is removed now:

* Replying in timeouts
* Subscribe and unsubscribe in handlers

Of course you can still get IBus injected to achieve it

danielmarbach commented:
@Particular/engineers discussed with @johnsimons that subscribing and unsubscribing as a startup and shutdown concern. Therefore, I only allow to do it with the `IRunContext`.

danielmarbach commented:
@SimonCropp would you mind taking a look at it as well before I pull it in tomorrow?

SimonCropp commented on src/NServiceBus.Core/Unicast/Behaviors/MessageHandler.cs:
```diff
@@ -7,14 +7,75 @@
+        /// <summary>
+        /// Creates a new instance of the message handler
+        /// </summary>
+        [ObsoleteEx(TreatAsErrorFromVersion = "6", RemoveInVersion = "7", ReplacementTypeOrMember = "MessageHandler(Action<object, object, object> invocationWithContext, Type handlerType)")]
```
you cant have an obsolete on a public API that points to a internal member as the replacement

danielmarbach replied:
What to do in this case?

SimonCropp replied:
if there is no valid replacement then just dont set ReplacementTypeOrMember

danielmarbach replied:
what do we do when users want to test a behavior which is using the MessageHandler class. If I don't expose the `invocationWithContext` delegate that class is not usable by end users without the real pipeline. Could that be a problem?

SimonCropp replied:
do we use it for unit testing?

SimonCropp commented on src/NServiceBus.AcceptanceTests/Exceptions/Handler_and_UowEnd_throws.cs:
```diff
@@ -24,37 +24,11 @@ public void Should_receive_AggregateException_with_both_exceptions()
             Assert.AreEqual(typeof(EndException), context.InnerExceptionTwoType);
-
-
-            StackTraceAssert.StartsWith(
```
so the stack trace asserts are removed. I look forward to the eventual regression

danielmarbach replied:
Eventual regression in what?

SimonCropp replied:
we previously had a several bugs that corrupted the stack trace in various ways. hence the test

SimonCropp commented on src/NServiceBus.Core/Unicast/Behaviors/MessageHandler.cs:
```diff
@@ -7,14 +7,75 @@
+    /// <summary>
+    /// TODO: Should we expose that to public?
+    /// </summary>
+    internal enum HandlerKind
```
moving this to its own file will make it more discoverable in the solution exploirer

danielmarbach replied:
Good point I forgot that

SimonCropp commented on src/NServiceBus.Core/Saga/TimeoutContext.cs:
```diff
@@ -0,0 +1,44 @@
+{
+    using System;
+
+    internal class TimeoutContext : ITimeoutContext
```
why not a base class?

danielmarbach replied:
Discussed this with @johnsimons and @andreasohlund we wanted to go towards a flatter structure without inheritance like we previously had with the send options (which is now changed as well). Let's address this in another PR ok?

andreasohlund commented:
Closing since we decided to start fresh since we don't need to support the old api
